### ISSUE TITLE
keystone should not be sensitive to module.parent

### ISSUE THREAD

**Issue by indexzero** [keystonejs/keystone#652]
This is a problem we encountered recently:

**bin/some-file.js**
``` js
var Schema = require('../schemas/some-schema'),
      keystone = require('keystone');

// Where there is a ../models directory
keystone.import('../models');
```

**schemas/some-schema.js**
``` js
var keystone = require('keystone');

/* Other code is irrelevant */
```

This caused the `keystone` module to have a parent that was different than our other main `index.js` script which because of [this code](https://github.com/keystonejs/keystone/blob/master/index.js#L14-L24) made many `keystone` helpers impossible to use until we reordered the requires.

This is a very subtle and very important bug because most novices will have absolutely no idea what is going on. You should not be using `module.parent`, but instead have some way to configure this that is required on setup to be  __the project root.__ Otherwise all of your requires can be borked for no reason just because of a different `require` ordering.

**Comment by morenoh149** [comment]
can you propose a solution?

**Comment by indexzero** [comment]
It's hard to say 100%, because I'm not everywhere that the `moduleRoot` variable it is being used as reference for is used. Basically you want the root of the module that you are being required in. 

_If you assume that keystone is always a first-level dependency then you can simply recurse up the directory tree to find it._

Honestly though I would just make a breaking change where you pass in the root into a `keystone.configure` method or something like that. Relying on any sort of internals of the `node` module system is almost always going to cause adverse side-effects.

**Comment by JohnnyEstilles** [comment]
Keystone uses `moduleRoot` for a number of functions, including finding the `views` path, the `favicon` path, etc. Most Keystone developers (myself included) `require('keystone')` somewhere in the root of the app, so this isn't an issue for the majority of us.

That said, I propose we do two things:
1. Document the fact that `Keystone` expects to get required at the root of the app (which should be the standard convention)
2. Add an optional setting to Keystone (e.g. `keystone.set('moduleRoot', '../path/to/root')` for those who wish to deviate from the standard convention

If we do decide to add the `moduleRoot` setting, we also need to make sure we document well that it needs to be set with a *relative* path (that would be expanded by Keystone). Otherwise the app could run into issues when being deployed (or simply if it's moved it to a different system or folder).

@alanshaw (issue #943), I typically separate the Keystone app instantiation/configuration and the startup into two separate files. That way I could require the file that instantiates Keystone from both the app entry point (which then starts the Keystone app) and a test helper that I then require in all my tests. This way Keystone is initially required form the app root, avoiding the moduleRoot issue. (I ran into this issue when I first started testing my Keystone apps).

@JedWatson, if you agree with my assessment above, I will submit a PR adding the optional `moduleRoot` setting.

**Comment by alanshaw** [comment]
Hey @JohnnyEstilles - thanks for the insight.

FWIW:

The fact that I can require keystone from different directories and get different behaviour is a code smell. I understand it's built like that for convenience for the majority of people but it should be, at the very least, configurable.

If I have to restructure my app to allow it to be tested then that's probably something that needs to be changed in the yeoman generator. We should break down any barriers in the way of people testing their code!

**Comment by JohnnyEstilles** [comment]
@alanshaw, I personally don't consider this a code small, but that's just my opinion. I sincerely do not believe this is a symptom of a deeper issue in our code base, IMHO. 

Suggesting that making the minor changes I suggested in order to make your code testable is "restructuring your app" is a bit of an overstatement. I was just trying to give you a temporary solution while we implement a more permanent one. 

The changes I'm proposing will make Keystone "configurable" as per your suggestion. Please bear with us while we make the required changes. 

I do agree that we should update the yeoman generator to make the generated Keystone app testable out of the box. 

I will be submitting PR with suggested corrections for both of these issues within then next couple of days (unless someone beats me to it). :-)

**Comment by SharadKumar** [comment]
+1

**Comment by JedWatson** [comment]
Fixed by #954
